asciidoc/courses/genai-mcp-build-custom-tools-python/modules/2-database-features/lessons/1-connecting-neo4j/lesson.adoc

```diff
@@ -0,0 +1,260 @@
+= Connecting to Neo4j
+:type: lesson
+:order: 1
+
+
+Now that you have a working MCP server, it's time to connect it to a database.
+
+In this lesson, you will learn how to connect your MCP server to Neo4j using FastMCP's lifespan management feature to properly handle database connections.
+
+
+== The Problem with Simple Connections
+
+Consider this naive approach to connecting to Neo4j:
+
+[source,python]
+----
+from neo4j import GraphDatabase
+
+@mcp.tool()
+def get_movies() -> list[dict]:
+    """Get a list of movies"""
+    # Creating a new driver for every tool call!
+    driver = GraphDatabase.driver(uri, auth=(user, password))
+    with driver.session() as session:
+        result = session.run("MATCH (m:Movie) RETURN m LIMIT 10")
+        return [record.data() for record in result]
+    driver.close()
+----
+
+This approach has several problems:
+
+* **Performance**: Creating a new driver connection for every tool call is slow and inefficient
+* **Resource leaks**: If the tool fails, the driver may not be closed properly
+* **Connection pooling**: Neo4j drivers maintain connection pools that should be reused across requests
+* **Best practices**: The Neo4j driver should be created once and closed when the server shuts down
+
+
+== Introducing Lifespan Management
+
+FastMCP provides a **lifespan** feature that allows you to:
+
+1. **Initialize resources** when the server starts (e.g., create database connections)
+2. **Clean up resources** when the server shuts down (e.g., close connections)
+3. **Share resources** across all tools and resources in your server
+
+
+=== The Lifespan Context Manager
+
+To share objects across the server, we can create a function that initializes resources when the server starts and cleans them up when it shuts down.
+
+The function yields an object that contains the FastMCP server will make available to any tools and resources that request it.
```

Comment:
```suggestion
The function yields an object that contains the FastMCP server that it will make available to any tools and resources that request it.
```